zypp/VendorAttr.cc

```diff
@@ -99,86 +94,83 @@ namespace zypp
-	  for ( VendorMap::reverse_iterator it = _vendorMap.rbegin(); it != _vendorMap.rend(); ++it )
+	  for ( VendorGroupMap::const_reverse_iterator it = _vendorGroupMap.rbegin(); it != _vendorGroupMap.rend(); ++it )
 	  {
-	    if ( str::hasPrefix( lcvendor.c_str(), it->first ) )
-	    {
+	    if ( str::hasPrefix( lcvendor.c_str(), it->first ) ) {
 	      myid = it->second;
 	      break; // found
 	    }
 	  }
 	}
+
 	if ( ! myid )
-	{
 	  myid = --_nextId; // get a new class ID
-	}
+
 	ent = lcent = myid; // remember the new DI
       }
       else
-      {
 	ent = lcent; // take the ID from the lowercased vendor string
-      }
     }
     return ent;
   }
 
-  void VendorAttr::Impl::addVendorList( VendorList && vendorList_rr )
+  void VendorAttr::Impl::addVendorList( VendorList && vendorList_r )
   {
-#warning REIMPLEMENT
-    std::vector<std::string> vendorList_r;
-    for ( const auto & el : vendorList_rr ) {
-      vendorList_r.push_back( el.asString() );
-    }
-
-    unsigned int nextId = vendorGroupCounter + 1;
-    // convert to lowercase and check if a vendor is already defined
-    // in an existing group.
+    // Will add a new equivalence group unless we merge with existing groups.
+    unsigned targetId = _vendorGroupId + 1;
 
-    for_( it, vendorList_r.begin(), vendorList_r.end() )
+    // (!) Convert the group strings in place to lowercase before adding/checking.
+    // Extend/join already existing groups if they are referenced.
+    for ( std::string & vendor : vendorList_r )
     {
-      *it = str::toLower( *it );
-      if (_vendorMap.find(*it) != _vendorMap.end())
+      vendor = str::toLower( std::move(vendor) );
+
+      if ( _vendorGroupMap.count( vendor ) )
       {
-        if (nextId != vendorGroupCounter + 1 &&
-            nextId != _vendorMap[*it])
-        {
-	  // We have at least 3 groups which has to be mixed --> mix the third group to the first
-          unsigned int moveID = _vendorMap[*it];
-          for_( itMap, _vendorMap.begin(), _vendorMap.end() )
-          {
-            if (itMap->second == moveID)
-              itMap->second = nextId;
-          }
-        }
-        else
-        {
-          nextId = _vendorMap[*it];
-          WAR << "Vendor " << *it << " is already used in another vendor group. --> mixing these groups" << endl;
-        }
+	unsigned joinId = _vendorGroupMap[vendor];
+	if ( targetId == _vendorGroupId + 1 ) {
+	  targetId = joinId;	// will extend the existing group
+	}
+	else if ( targetId != joinId ) {
+	  // yet another existing group -> join it into the target group
+	  for ( auto & el : _vendorGroupMap ) {
+	    if ( el.second == joinId )
+	      el.second = targetId;
+	  }
+	}
+	vendor.clear();	// no need to add it later
       }
     }
-	// add new entries
-    for_( it, vendorList_r.begin(), vendorList_r.end() )
-    {
-      _vendorMap[*it] = nextId;
+
+    // Now add the new entries
+    for ( std::string & vendor : vendorList_r ) {
+      if ( ! vendor.empty() )
```

Comment:
No need to go over the list if vendor is empty, so the if should enclose the for loop

Reply:
You mixed `std::string & vendor` and `vendorList_r`, didn't you? We just skip empty strings in the list.

Reply:
Right I did not realize the vendor variable is shadowed here, sorry